### DISCUSSION TITLE
Use NN.Conv without edge_attr

### DISCUSSION THREAD

**Discussion by AdarshMJ** [pyg-team/pytorch_geometric#4561] [Q&A] [answered]
Hi! Im trying to use 
`from torch_geometric.nn import NNConv` 
The documentation says -
 nn ([torch.nn.Module](https://pytorch.org/docs/master/generated/torch.nn.Module.html#torch.nn.Module)) – A neural network 
 that maps edge features edge_attr of shape [-1, num_edge_features] to shape [-1, in_channels * out_channels]

 The dataset Im working with doesn't have edge_attr, is there a way to use this without edge_attr? Any ideas are appreciated! Thank you.

**Comment by rusty1s** [comment, marked as answer]
No, `NNConv` is defined to operate on edge features, and so edge features will be required to run this operator. As a workaround, you could think of creating edge features on the fly, e.g., by concatenating source and destination node features:
```python
edge_attr = torch.cat([x[edge_index[0]], x[edge_index[1]], dim=-1)
```

**Reply by AdarshMJ** [reply]
Thank you for the suggestion, I will try this!

**Reply by AdarshMJ** [reply]
Hi Matthias,
So I think this works, I have created the edge_attrs. Thank you again.
I have two questions slightly unrelated -


1)After incorporating the edge_attrs, I am getting an error - `RuntimeError: "addmm_cuda" not implemented for 'Long'` Pytorch has this issue highlighted here - https://github.com/pytorch/pytorch/issues/44428
Im not sure what part is giving this error. Should I just convert all my tensors to float64 or something?

2)In PNA, how do you set the `self.edge_emb = Embedding()` My data looks like this. -
Number of graphs in the current batch: 64
DataBatch(x=[559, 1], edge_index=[2, 4911], y=[9600], edge_attr=[2388], batch=[559], ptr=[65])
Number of graphs in the current batch: 64
DataBatch(x=[562, 1], edge_index=[2, 4950], y=[9600], edge_attr=[2448], batch=[562], ptr=[65])

Thank you for your time.

**Reply by rusty1s** [reply]
1) All features need to be of type floating-point (at best `float32`). The error suggests that this is not the case for you.
2) Note that you only need `self.edge_emb = Embedding()` in case your `edge_attr` holds integer values. You can use the `Embedding()` layer to map them to floating-point types:
```python
edge_emb = Embedding(num_categories, out_channels)
edge_attr = edge_emb(edge_attr)
```

**Reply by AdarshMJ** [reply]
Hi @rusty1s Took me some time, but it works. Thank you! 
For creating the edge_attr, I am currently doing -

```
def transforms(data):
edge_attr = torch.cat([data.edge_index[0],data.edge_index[1]], dim=-1)
```
and then applying it to my dataset as -
`dataset = [transforms(data) for data in dataset]`

But the edge_attr created are of the form -

```
Number of graphs in the current batch: 64
DataBatch(x=[386, 1], edge_index=[2, 2364], y=[9600], edge_attr=[1488], batch=[386], ptr=[65])

Step 2:
=======
Number of graphs in the current batch: 64
DataBatch(x=[389, 1], edge_index=[2, 2415], y=[9600], edge_attr=[1500], batch=[389], ptr=[65])
```
Is there a way to create them in the shape ?
`data.edge_attr: Edge feature matrix with shape [num_edges, num_edge_features]`

Thank you in advance for your time :)

**Reply by rusty1s** [reply]
```python
def transforms(data):
    edge_attr = torch.cat([data.x[data.edge_index[0]], data.x[data.edge_index[1]]], dim=-1)
```
should fix this.

**Reply by AdarshMJ** [reply]
Thank you!
